job-manager: support mode=limited

chu11 opened the pull request:
job-manager / sched-simple: Support mode=N pending jobs depth

Per #2070, support `mode=N` option, where N is the number of pending jobs that can be concurrently sent to the scheduler.  Apologies to @grondo on any conflicts with #3464.

- I have a RFC27 change here: https://github.com/flux-framework/rfc/pull/284.  Brought up some questions there about how to add `mode=N`.  I'll leave some discussions over there.

Notes specific to implementation here:

- for consistency, `mode` in `sched-simple` is now `alloc-mode`.  Added `mode` in `sched-simple` to be `job-manager` mode (`single`, `unlimited`, etc.).

- broke backwards compatibility by removing `sched-simple`'s "unlimited" option.  I made it `mode=unlimited` now.  Should add back as legacy option?

- `sched-simple` defaulted to `single` before, even if the user inputted a bad options.  I left that behavior, but perhaps this should be changed.

- right now the alloc queue and pending_jobs lists are implemented with `zlistx`.  Longer term, we may want to look into another data structure, such as heap, to reduce insertion time. (will create an issue later, issue also exists in `sched-simple`).

- after this PR, we really whould re-organize / cleanup the organization of the job-manager tests (#3417).  Maybe check for duplicate tests or combine tests, or make some helper functions, etc.

Edit:

Oh yeah.  I added a `pending_jobs` list in `job-manager`.  For the purposes of this PR I only needed the minimum priority job, so the ordered list is a bit overkill (I only grab the tail of the list).  I did the list under the assumption we would need it in the future after priority plugin.  Perhaps I should just store the minimum and leave the `pending_jobs` list to PR #3464?

garlick commented:
@chu11, well thanks for attacking a long standing problem.  Hmm, there are a few open questions here, and we have the pending priority plugin work in the job manager.  Seems like this should probably be deferred until after we get priority plugin PR merged, and we've gotten our monthly release done?

chu11 commented:
> @chu11, well thanks for attacking a long standing problem. Hmm, there are a few open questions here, and we have the pending priority plugin work in the job manager. Seems like this should probably be deferred until after we get priority plugin PR merged, and we've gotten our monthly release done?

Sounds good, didn't know if we wanted to sneak in before release and have priority plugin PR for next release, or vice versa.

grondo commented:
This issue I was thinking a `limited` mode solves is with an active priority plugin which may reprioritize all jobs. When there is a large queue the job-manager has to update the scheduler about *all* jobs (with `unlimited` mode). With a `limited` mode, the impact of a periodic reprioritization loop could be mitigated by limiting the number of jobs for which the scheduler needs to be notified. Though, actually now that I think about it, this mitigation could be offset by an increased need for alloc cancellation RPCs, hmm.

In any event, support for a limited mode isn't really a critical issue until after the priority plugin work is implemented, and a potential for reprioritization is added.

chu11 commented:
rebased & re-pushed a WIP drop:

- updated per RFC27 changes
- built on top of PR #3464 

TODO: mostly tests to ensure mode=limit works when re-prioritizing > 1 job (or all jobs) at a time.  I imagine once #3464 tests are written I could piggy back off of them.

chu11 commented:
rebased on master.

- created a new plugin for testing called `priority-invert`, which will invert all of the job prioties (i.e. 1 turns to 30, 30 turns to 1).  It's mostly for testing to ensure that all pending jobs can be cancelled in the scheduler and replaced with (new high priority) queued jobs.

- added a number of tests for the `mode=limited` cases.

grondo commented on src/modules/job-manager/alloc.c:
```diff
@@ -579,13 +609,22 @@ int alloc_queue_reprioritize (struct alloc *alloc)
     }
 
-    if (alloc->mode == SCHED_SINGLE)
+    zlistx_sort (alloc->pending_jobs);
+
+    job = zlistx_first (alloc->pending_jobs);
+    while (job) {
+        job->handle = zlistx_cursor (alloc->pending_jobs);
+        job = zlistx_next (alloc->pending_jobs);
+    }
+
+    if (alloc->alloc_pending_max)
```
All this code is duplicated from above -- possibly happened during a rebase?

grondo replied:
Oops, sorry! Didn't notice these are different lists at first! D'oh!

chu11 commented:
rebased & re-pushed, fixing up @grondo's comment above.  Went ahead and squashed commits b/c it's just a search & replace.  Also fixed up a chain-lint bug.

Removed WIP, as I think this is close.

chu11 commented:
doh! i had forgotten that `flux-sched` needs to be updated for these changes.  I should probably maintain backwards compatibility of the old config style for  `flux-sched`, then update `flux-sched`, then come back to `flux-core` and remove the legacy config style.

dongahn commented:
> Edit: nevermind, flux-sched should be compatible to this. I guess the builder failed for a different reason.

> FAIL: t1018-rv1-bootstrap2.t 12 - rv1-bootstrap2: 2N nesting works (policy=low)

Yes and this should be looked at but it didn't bubble up to the top of my queue. Go ahead and restart the failed job.

grondo commented:
This error message could be improved:
```
$ flux module reload sched-simple mode=limite=5
err: unknown mode: limite=5
: Bad file descriptor
```

chu11 commented:
> This error message could be improved:

Doh, using `flux_log_error()` was a cut & paste from the error message for `get_alloc_mode()`.  I'll fix it there too.

chu11 commented:
re-pushed, fixing up the error message @grondo found above, plus the other location with a similar error message issue.  Also renamed `alloc_concurrency` variable to `alloc_limit`, for consistency.  Everything is already squashed, since the changes are so minor.

grondo commented:
Just fyi - ran the following suite of tests on this branch with sched-simple:

`test.sh`:
```sh
#!/bin/sh
echo "Testing with mode=$MODE"
flux module reload -f sched-simple mode=$MODE
./throughput.py -n 1024
```

```
 grondo@asp:~/git/fc.git$ for i in 1 2 3 4 8 16 32 64 128 256 1024; do echo "mode=limited=$i"; MODE=limited=$i src/cmd/flux start -s 8 sh ./test.sh; done; MODE="unlimited" src/cmd/flux start -s 8 sh test.sh
mode=limited=1
Testing with mode=limited=1
number of jobs: 1024
script runtime: 11.284s
job runtime:    11.076s
throughput:     92.5 job/s (script:  90.7 job/s)
mode=limited=2
Testing with mode=limited=2
number of jobs: 1024
script runtime: 8.903 s
job runtime:    8.720 s
throughput:     117.4 job/s (script: 115.0 job/s)
mode=limited=3
Testing with mode=limited=3
number of jobs: 1024
script runtime: 8.048 s
job runtime:    7.864 s
throughput:     130.2 job/s (script: 127.2 job/s)
mode=limited=4
Testing with mode=limited=4
number of jobs: 1024
script runtime: 7.397 s
job runtime:    7.208 s
throughput:     142.1 job/s (script: 138.4 job/s)
mode=limited=8
Testing with mode=limited=8
number of jobs: 1024
script runtime: 7.331 s
job runtime:    7.147 s
throughput:     143.3 job/s (script: 139.7 job/s)
mode=limited=16
Testing with mode=limited=16
number of jobs: 1024
script runtime: 7.186 s
job runtime:    6.999 s
throughput:     146.3 job/s (script: 142.5 job/s)
mode=limited=32
Testing with mode=limited=32
number of jobs: 1024
script runtime: 7.123 s
job runtime:    6.931 s
throughput:     147.7 job/s (script: 143.8 job/s)
mode=limited=64
Testing with mode=limited=64
number of jobs: 1024
script runtime: 7.095 s
job runtime:    6.886 s
throughput:     148.7 job/s (script: 144.3 job/s)
mode=limited=128
Testing with mode=limited=128
number of jobs: 1024
script runtime: 7.185 s
job runtime:    6.997 s
throughput:     146.4 job/s (script: 142.5 job/s)
mode=limited=256
Testing with mode=limited=256
number of jobs: 1024
script runtime: 7.350 s
job runtime:    7.144 s
throughput:     143.3 job/s (script: 139.3 job/s)
mode=limited=1024
Testing with mode=limited=1024
number of jobs: 1024
script runtime: 7.461 s
job runtime:    7.272 s
throughput:     140.8 job/s (script: 137.3 job/s)
Testing with mode=unlimited
number of jobs: 1024
script runtime: 7.074 s
job runtime:    6.891 s
throughput:     148.6 job/s (script: 144.7 job/s)
```

I wonder if we should set mode=limited=8 or 32 in sched-simple by default.

chu11 commented:
re-pushed with an extra commit on top to set the default limit to 8.  Had to update some tests that assumed limit of 1, but that was it otherwise.  And I had forgotten to add a "Fixes ..." into one of the commit messages.  Added it to one of the commits.

grondo reviewed (approved):
LGTM!

Testing this out extensively and found no issues. Thanks!

chu11 commented:
@grondo thanks!  I remember @garlick saying it wanted to take a skim at this too.  So i'll wait to set MWP if he has any thoughts.

garlick commented:
> I remember @garlick saying it wanted to take a skim at this too. So i'll wait to set MWP if he has any thoughts.

I'm out through tues so if you guys are ready, I say go for it.  This is a nice addition!

chu11 commented:
ok, setting MWP
